### PR TITLE
aws_lakeformation_permissions support AllIAMPrincipals special principal

### LINKED ISSUE #29767: [Bug]: aws_lakeformation_permissions does not support IAMPrincipals
### Terraform Core Version

1.1.8

### AWS Provider Version

4.56.0

### Affected Resource(s)

aws_lakeformation_permissions

### Expected Behavior

`DESCRIBE` permissions granted to all Principals in the account via: `principal = "123456789012:IAMPrincipals" `

### Actual Behavior

A Terraform error is thrown

### Relevant Error/Panic Output Snippet

`Error: "principal" doesn't look like AWS Account ID (exactly 12 digits): "123456789012:IAMPrincipals"`

### Terraform Configuration Files

none to add

### Steps to Reproduce

```hcl
data "aws_caller_identity" "current" {}

resource "aws_glue_catalog_database" "database" {
  name  = "my_test_db"
}

resource "aws_lakeformation_permissions" "local_account_describe" {
  depends_on = [
    aws_glue_catalog_database.database
  ]
  principal   = "${data.aws_caller_identity.current.id}:IAMPrincipals" # <-- this
  permissions = ["DESCRIBE"]
  database {
    name       = aws_glue_catalog_database.database.name
    catalog_id = data.aws_caller_identity.current.id
  }
}

```

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

- https://docs.aws.amazon.com/lake-formation/latest/dg/granting-db-perms-api.html
- https://docs.aws.amazon.com/lake-formation/latest/dg/granting-location-permissions-external.html
- https://docs.aws.amazon.com/lake-formation/latest/dg/granting-location-permissions-local.html

Each of the 3 Official documentation pages describe an account-level grant that's possible through `123456789012:IAMPrincipals` but using this approach fails with terraform with the provided error message.

A sample `aws` CLI command that works as expected:

```bash
aws lakeformation grant-permissions \
    --principal DataLakePrincipalIdentifier='111122223333:IAMPrincipals' \
    --permissions DESCRIBE \
    --resource '{"Database": {"Name": "my_test_db"} }'
```

### Would you like to implement a fix?

No

### PR DESCRIPTION
**Pull request by nickdelnano**
Closes #29767

Lake Formation supports two special principal values that are defined by the Lake Formation service:
- `IAMAllowedPrincipals` (already supported by `aws_lakeformation_permissions` resource since its [creation](https://github.com/hashicorp/terraform-provider-aws/pull/13396/files#diff-89223e5e24bbc2ac9c3ac2a3e676a5b29529b8ab86def7f803d200bd4d3e1728R370))
- `AllIAMPrincipals` (**Added in this PR**)

The implementation and test cases for `AllIAMPrincipals` closely follows that of `IAMAllowedPrincipals`.

> Principal is an IAM group - ALLIAMPrincipals
> When you grant permissions to ALLIAMPrincipals group on a Data Catalog resource, every principal in the account gets 
> access to the Data Catalog resource using Lake Formation permissions and IAM permissions.
> Example:
>> --principal DataLakePrincipalIdentifier=123456789012:IAMPrincipals

https://docs.aws.amazon.com/lake-formation/latest/dg/lf-permissions-reference.html

### CODE REVIEW

File: website/docs/r/lakeformation_permissions.html.markdown [hunk 1]
```diff
@@ -87,6 +87,21 @@ The resulting permissions depend on whether the table had `IAMAllowedPrincipals`
...
+  permissions = ["SELECT"]
+  principal   = "123456789012:IAMPrincipals"
+
+  table_with_columns {
+    database_name = aws_glue_catalog_table.example.database_name
+    name          = aws_glue_catalog_table.example.name
+    column_names  = ["event"]
+  }
+}
+
```
Comment (nickdelnano) [inline website/docs/r/lakeformation_permissions.html.markdown:104]: https://docs.aws.amazon.com/lake-formation/latest/dg/lf-permissions-reference.html

File: internal/service/lakeformation/validate_test.go [hunk 2]
```diff
@@ -42,7 +43,9 @@ func TestValidPrincipal(t *testing.T) {
 	invalidNames := []string{
 		"IAM_NOT_ALLOWED_PRINCIPALS", // doesn't exist
 		names.AttrARN,
-		"1234567890125", //not an account id
+		"1234567890125",               //not an account id
+		"IAMPrincipals",               // incorrect representation
+		"1234567890125:IAMPrincipals", // incorrect representation, account id invalid length
```
Comment (nickdelnano) [inline internal/service/lakeformation/validate_test.go:48]: best viewed with "hide whitespace" enabled

File: internal/service/lakeformation/lakeformation_test.go [hunk 3]
```diff
@@ -54,6 +55,7 @@ func TestAccLakeFormation_serial(t *testing.T) {
 		"PermissionsTable": {
 			acctest.CtBasic:      testAccPermissions_tableBasic,
 			"iamAllowed":         testAccPermissions_tableIAMAllowed,
+			"iamPrincipals":      testAccPermissions_tableIAMPrincipals,
```
Comment (nickdelnano) [inline internal/service/lakeformation/lakeformation_test.go:58]: best viewed with "hide whitespace" enabled

**Comment by nickdelnano** [comment]
Hi @justinretzolk , do you know when this change could be reviewed and hopefully merged? Its a small change but meaningfully adds support for a missing primitive in AWS Lake Formation permissions.

Thank you for your time! I know there is a considerable workload in reviewing PRs.

**Review by johnsonaj — APPROVED** [review]
LGTM 🚀 

```console
% go test -count=1 -v ./internal/service/lakeformation/... -run='TestValidPrincipal’

2024/12/18 09:34:51 Initializing Terraform AWS Provider...
=== RUN   TestValidPrincipal
=== PAUSE TestValidPrincipal
=== CONT  TestValidPrincipal
--- PASS: TestValidPrincipal (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/lakeformation	10.935s
```

```console
% make testacc TESTARGS="-run=TestAccLakeFormation_serial" PKG=lakeformation

make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.2 test ./internal/service/lakeformation/... -v -count 1 -parallel 20  -run=TestAccLakeFormation_serial -timeout 360m
2024/12/17 14:02:46 Initializing Terraform AWS Provider...
--- PASS: TestAccLakeFormation_serial (1408.45s)
    --- PASS: TestAccLakeFormation_serial/PermissionsTableWithColumns (149.30s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTableWithColumns/wildcardSelectPlus (25.40s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTableWithColumns/basic (55.89s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTableWithColumns/implicit (22.18s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTableWithColumns/wildcardExcludedColumns (22.93s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTableWithColumns/wildcardSelectOnly (22.90s)
    --- PASS: TestAccLakeFormation_serial/LFTags (100.30s)
        --- PASS: TestAccLakeFormation_serial/LFTags/basic (15.53s)
        --- PASS: TestAccLakeFormation_serial/LFTags/disappears (12.73s)
        --- PASS: TestAccLakeFormation_serial/LFTags/tagKeyComplex (12.02s)
        --- PASS: TestAccLakeFormation_serial/LFTags/values (25.72s)
        --- PASS: TestAccLakeFormation_serial/LFTags/valuesOverFifty (34.30s)
    --- PASS: TestAccLakeFormation_serial/ResourceLFTag (59.45s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTag/basic (14.02s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTag/disappears (15.21s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTag/table (15.34s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTag/tableWithColumns (14.89s)
    --- PASS: TestAccLakeFormation_serial/DataCellsFilter (74.82s)
        --- PASS: TestAccLakeFormation_serial/DataCellsFilter/basic (18.35s)
        --- PASS: TestAccLakeFormation_serial/DataCellsFilter/columnWildcard (14.99s)
        --- PASS: TestAccLakeFormation_serial/DataCellsFilter/disappears (16.02s)
        --- PASS: TestAccLakeFormation_serial/DataCellsFilter/rowFilter (25.45s)
    --- PASS: TestAccLakeFormation_serial/PermissionsBasic (317.72s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/basic (21.50s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/database (22.18s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/databaseIAMAllowed (43.58s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/databaseIAMPrincipals (13.93s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/dataLocation (20.17s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/disappears (82.23s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/lfTag (22.50s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/lfTagPolicyMultiple (22.07s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/databaseMultiple (22.51s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/dataCellsFilter (24.52s)
        --- PASS: TestAccLakeFormation_serial/PermissionsBasic/lfTagPolicy (22.53s)
    --- PASS: TestAccLakeFormation_serial/PermissionsTable (248.17s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/wildcardNoSelect (22.15s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/wildcardSelectPlus (23.46s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/iamAllowed (45.09s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/iamPrincipals (15.21s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/implicit (23.59s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/multipleRoles (24.12s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/basic (23.58s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/selectOnly (23.71s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/selectPlus (23.62s)
        --- PASS: TestAccLakeFormation_serial/PermissionsTable/wildcardSelectOnly (23.64s)
    --- PASS: TestAccLakeFormation_serial/ResourceLFTags (163.59s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTags/disappears (14.55s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTags/hierarchy (30.53s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTags/table (26.80s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTags/tableWithColumns (27.02s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTags/basic (14.52s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTags/database (25.50s)
        --- PASS: TestAccLakeFormation_serial/ResourceLFTags/databaseMultipleTags (24.67s)
    --- PASS: TestAccLakeFormation_serial/DataLakeSettings (85.26s)
        --- PASS: TestAccLakeFormation_serial/DataLakeSettings/basic (12.12s)
        --- PASS: TestAccLakeFormation_serial/DataLakeSettings/disappears (13.20s)
        --- PASS: TestAccLakeFormation_serial/DataLakeSettings/withoutCatalogId (12.52s)
        --- PASS: TestAccLakeFormation_serial/DataLakeSettings/readOnlyAdmins (12.02s)
        --- PASS: TestAccLakeFormation_serial/DataLakeSettings/parameters (35.39s)
    --- PASS: TestAccLakeFormation_serial/DataLakeSettingsDataSource (25.53s)
        --- PASS: TestAccLakeFormation_serial/DataLakeSettingsDataSource/readOnlyAdmins (12.95s)
        --- PASS: TestAccLakeFormation_serial/DataLakeSettingsDataSource/basic (12.58s)
    --- PASS: TestAccLakeFormation_serial/PermissionsDataSource (184.32s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/basic (22.18s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/dataCellsFilter (24.78s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/database (23.18s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/dataLocation (21.77s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/lfTag (22.71s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/lfTagPolicy (23.04s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/table (22.40s)
        --- PASS: TestAccLakeFormation_serial/PermissionsDataSource/tableWithColumns (24.26s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/lakeformation	1414.830s
```

**Comment by johnsonaj** [comment]
@nickdelnano thank you for the contribution! 🎉

**Pull request merged**